### PR TITLE
ap-4122: add purge_worker job

### PR DESCRIPTION
**Pull request by kmahern**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/AP-4122)

Added purge job to run nightly at 8pm
Purge bulk submission original and result files
Anonymise associated submission records

## Checklist

Before you ask people to review this PR:

- Tests and linters should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: spec/workers/purge_worker_spec.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  describe "#perform" do
+    subject(:perform) { described_class.new.perform }
+
+    let(:bulk_submission) { create(:bulk_submission, :with_original_file, :with_result_file, expires_at:) }
+    let(:submission) do
+ create(:submission, first_name: 'Rosie', last_name: 'Conway', nino: 'JC654321A', dob: '1977-03-08'.to_date,  
+bulk_submission:) end
+    let(:expires_at) { Time.zone.now }
```
Comment (jsugarman) [inline spec/workers/purge_worker_spec.rb:31]: **nit** `Time.current` or, to more close reflect reality, `1.day.ago.midnight` - this is similar to what `expires_at` is set too

File: spec/workers/purge_worker_spec.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    before do
+      bulk_submission
+      submission
+    end
+
+    it_behaves_like "applcation worker logger"
+
+    context "when the bulk submission has expired" do
```
Comment (jsugarman) [inline spec/workers/purge_worker_spec.rb:40]: makes more sense to define the test setup logic for expiry here

```
context "when the bulk submission has expired" do
  let(:expires_at) { 1.day.ago.midnight }
  ...
```

File: spec/workers/purge_worker_spec.rb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      it "purges the original file" do
+        perform
+        bulk_submission.reload
+        expect(bulk_submission.original_file.attachment).to be_nil
+        expect(bulk_submission.original_file.blob).to be_nil
```
Comment (jsugarman) [inline spec/workers/purge_worker_spec.rb:45]: **nit** can use [compound expectations](https://makandracards.com/makandra/58102-rspec-3-allows-chaining-multiple-expectations) here

File: spec/workers/purge_worker_spec.rb [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        perform
+        bulk_submission.reload
+        expect(bulk_submission.result_file.attachment).not_to be_nil
+        expect(bulk_submission.result_file.blob).not_to be_nil
+      end
+
+      it "does not update the submission record" do
+        perform
+        submission.reload
+        expect(submission.first_name).to eq 'Rosie'
```
Comment (jsugarman) [inline spec/workers/purge_worker_spec.rb:85]: **nit** compound expectation or have_attributes as above

File: spec/workers/purge_worker_spec.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        submission.reload
+        expect(submission.first_name).to eq 'purged'
+        expect(submission.last_name).to eq 'purged'
+        expect(submission.nino).to eq 'AB123456C'
+        expect(submission.dob).to eq '01-01-1970'
+      end
+    end
+
+    context "when the bulk submission has not expired" do
+      let(:expires_at) { Time.zone.tomorrow }
```
Comment (jsugarman) [inline spec/workers/purge_worker_spec.rb:66]: We should probably test the boundaries with a life like test that freezes time at 8pm - or whenever it is scheduled - and then expects the one that has expires_at for previous day/midnight to be purged/cleaned/removed but one with expires_at today at midnight to no be purged.
```suggestion
      let(:expires_at_yesterday) { 1.day.ago.midnight }
      let(:expires_at_today) { Time.current.midnight }
```
I wonder whether we need to also change functionality and test to cover the very edges
```
      let(:expires_at_yesterday) { 1.day.ago.midnight }
      let(:expires_at_this_am) { 1.day.ago.midnight + 1.second }
      let(:expires_at_future) { Time.current.midnight }
```

if we expire something manually at 13:05 what do we expect it to do? (open question)

**Review by jsugarman — COMMENTED** [review]
looks good but a few comments and some things that should be changed
- must nullify `hmrc_interface_result`
- handle error scenarios where `expires_at` not set

File: app/workers/purge_worker.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+class PurgeWorker < ApplicationWorker
```
[the remark this answers is not in the document]
Comment (kmahern) [reply app/workers/purge_worker.rb]: Have renamed to PurgeSensitiveDataWorker

File: app/workers/purge_worker.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+  def perform
+    bulk_submissions = BulkSubmission.where(expires_at: ..Time.zone.now)
```
[the remark this answers is not in the document]
Comment (kmahern) [reply app/workers/purge_worker.rb]: 👍 have added purgeable_bulk_sumissions method

File: spec/workers/purge_worker_spec.rb (hunk 3 above)
Comment (kmahern) [reply spec/workers/purge_worker_spec.rb]: Have amended to use have_attributes to make one line expectation

File: spec/workers/purge_worker_spec.rb [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        expect(bulk_submission.result_file.blob).to be_nil
+      end
+
+      it "updates the submission record" do
+        perform
+        submission.reload
+        expect(submission.first_name).to eq 'purged'
+        expect(submission.last_name).to eq 'purged'
+        expect(submission.nino).to eq 'AB123456C'
+        expect(submission.dob).to eq '01-01-1970'
```
[the remark this answers is not in the document]
Comment (kmahern) [reply spec/workers/purge_worker_spec.rb]: 👍

File: spec/workers/purge_worker_spec.rb (hunk 5 above)
Comment (kmahern) [reply spec/workers/purge_worker_spec.rb]: see https://github.com/ministryofjustice/laa-assure-hmrc-data/pull/280/commits/394c7764ff6f4168b1b1f098362d5e6c303f5e7c

File: app/workers/purge_worker.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    bulk_submissions.each do |bulk_submission|
+      bulk_submission.original_file.purge
+      bulk_submission.result_file.purge
+      bulk_submission.submissions.each do |submission|
+        submission.update!(first_name: 'purged', last_name: 'purged', dob: Date.parse('1970-01-01'), nino: 'AB123456C')
```
[the remark this answers is not in the document]
Comment (kmahern) [reply app/workers/purge_worker.rb]: Thanks, setting to '{}' when puged

File: spec/workers/purge_sensitive_data_worker_spec.rb [hunk 10]
```diff
@@ -28,16 +28,21 @@
     let(:submission) do
  create(:submission, :with_completed_use_case_one_hmrc_interface_result, first_name: 'Rosie', last_name: 'Conway', 
 nino: 'JC654321A', dob: '1977-03-08'.to_date, bulk_submission:) end
-    let(:expires_at) { 1.day.ago.midnight }
+    let(:expires_at) { nil }
 
     before do
+      scheduled_time = Time.zone.local(2023, 0o6, 14, 20, 0, 0)
```
Comment (jsugarman) [inline spec/workers/purge_sensitive_data_worker_spec.rb:34]: i can see it works but what is the `0o6`. it seems to do the same thing with Time.zone.local(2023, 6, 14, 20, 0, 0).

**nit** Alternative which may be easier to read but either way
```
Time.current.midnight + 20.hours
```

Comment (jsugarman) [reply spec/workers/purge_sensitive_data_worker_spec.rb:34]: actually don't you need to fix the time to current date, rather than a fixed date?

**Review by jsugarman — APPROVED** [review]
Looks good 👍

**Pull request merged**